File: rholang/src/main/scala/coop/rchain/rholang/externalservices/OllamaService.scala
package coop.rchain.rholang.externalservices

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.model._
import akka.http.scaladsl.unmarshalling.Unmarshal
import akka.stream.Materializer
import cats.effect.{Concurrent, Sync}
import cats.syntax.all._
import com.typesafe.config.ConfigFactory
import com.typesafe.scalalogging.Logger
import spray.json._
import spray.json.DefaultJsonProtocol._

import java.util.concurrent.Executors
import scala.concurrent.{ExecutionContext, Future}
import scala.concurrent.duration._
import scala.util.control.NonFatal

// JSON protocol for Ollama API
object OllamaJsonProtocol extends DefaultJsonProtocol {
  case class OllamaMessage(role: String, content: String)
  case class OllamaChatRequest(
      model: String,
      messages: List[OllamaMessage],
      stream: Boolean = false
  )
  case class OllamaGenerateRequest(model: String, prompt: String, stream: Boolean = false)
  case class OllamaChatResponse(message: OllamaMessage, done: Boolean)
  case class OllamaGenerateResponse(response: String, done: Boolean)
  case class OllamaModel(name: String, modified_at: String, size: Long)
  case class OllamaModelsResponse(models: List[OllamaModel])

  implicit val messageFormat          = jsonFormat2(OllamaMessage)
  implicit val chatRequestFormat      = jsonFormat3(OllamaChatRequest)
  implicit val generateRequestFormat  = jsonFormat3(OllamaGenerateRequest)
  implicit val chatResponseFormat     = jsonFormat2(OllamaChatResponse)
  implicit val generateResponseFormat = jsonFormat2(OllamaGenerateResponse)
  implicit val modelFormat            = jsonFormat3(OllamaModel)
  implicit val modelsResponseFormat   = jsonFormat1(OllamaModelsResponse)
}

trait OllamaService {

  /** Chat completion using Ollama's chat endpoint */
  def chatCompletion[F[_]](model: String, prompt: String)(
      implicit F: Concurrent[F]
  ): F[String]

  /** Text generation using Ollama's generate endpoint */
  def textGeneration[F[_]](model: String, prompt: String)(
      implicit F: Concurrent[F]
  ): F[String]

  /** List available models from Ollama */
  def listModels[F[_]]()(
      implicit F: Concurrent[F]
  ): F[List[String]]
}

class DisabledOllamaService extends OllamaService {

  private[this] val logger: Logger = Logger[this.type]

  def chatCompletion[F[_]](model: String, prompt: String)(implicit F: Concurrent[F]): F[String] = {
    logger.debug("Ollama service is disabled - chatCompletion request ignored")
    F.raiseError(new UnsupportedOperationException("Ollama service is disabled via configuration"))
  }

  def textGeneration[F[_]](model: String, prompt: String)(implicit F: Concurrent[F]): F[String] = {
    logger.debug("Ollama service is disabled - textGeneration request ignored")
    F.raiseError(new UnsupportedOperationException("Ollama service is disabled via configuration"))
  }

  def listModels[F[_]]()(implicit F: Concurrent[F]): F[List[String]] = {
    logger.debug("Ollama service is disabled - listModels request ignored")
    F.raiseError(new UnsupportedOperationException("Ollama service is disabled via configuration"))
  }
}

class OllamaServiceImpl extends OllamaService {
  import OllamaJsonProtocol._

  private[this] val logger: Logger = Logger[this.type]

  implicit private val ec: ExecutionContext =
    ExecutionContext.fromExecutor(Executors.newCachedThreadPool())
  private val system                              = ActorSystem()
  implicit private val materializer: Materializer = Materializer(system)

  // Build Ollama client configuration
  private val config = ConfigFactory.load()
  private val baseUrl: String =
    if (config.hasPath("ollama.base-url")) config.getString("ollama.base-url")
    else "http://localhost:11434"

  private val defaultModel: String =
    if (config.hasPath("ollama.default-model")) config.getString("ollama.default-model")
    else "llama4:latest"

  private val timeoutSec: Int =
    if (config.hasPath("ollama.timeout-sec")) config.getInt("ollama.timeout-sec")
    else 30

  // Validate connection on startup
  validateConnectionOrFail()

  // shutdown system before jvm shutdown
  sys.addShutdownHook {
    system.terminate()
  }

  private def validateConnectionOrFail(): Unit = {
    val doValidate: Boolean =
      if (config.hasPath("ollama.validate-connection"))
        config.getBoolean("ollama.validate-connection")
      else true

    if (!doValidate) {
      logger.info(
        "Ollama connection validation is disabled by config 'ollama.validate-connection=false'"
      )
      return
    }

    try {
      // Test connection by trying to list models
      val future = Http()(system).singleRequest(
        HttpRequest(
          method = HttpMethods.GET,
          uri = s"$baseUrl/api/tags"
        )
      )

      val response = scala.concurrent.Await.result(future, timeoutSec.seconds)
      if (response.status.isSuccess()) {
        logger.info(s"Ollama service connection validated successfully at $baseUrl")
      } else {
        throw new RuntimeException(s"Ollama service responded with status: ${response.status}")
      }
    } catch {
      case NonFatal(e) =>
        throw new IllegalStateException(
          s"Ollama service connection validation failed. Check that Ollama is running on $baseUrl",
          e
        )
    }
  }

  def chatCompletion[F[_]](model: String, prompt: String)(
      implicit F: Concurrent[F]
  ): F[String] = {
    val actualModel   = if (model.isEmpty) defaultModel else model
    val promptPreview = if (prompt.length <= 250) prompt else prompt.take(250) + "..."
    logger.info(
      s"Ollama chatCompletion - input model: '$model', actualModel: '$actualModel', defaultModel: '$defaultModel', prompt: '$promptPreview'"
    )
    val requestBody = OllamaChatRequest(
      model = actualModel,
      messages = List(OllamaMessage("user", prompt)),
      stream = false
    ).toJson.toString()

    val future: Future[String] = for {
      response <- Http()(system).singleRequest(
                   HttpRequest(
                     method = HttpMethods.POST,
                     uri = s"$baseUrl/api/chat",
                     entity = HttpEntity(ContentTypes.`application/json`, requestBody)
                   )
                 )
      responseBody <- Unmarshal(response.entity).to[String]
      chatResponse = responseBody.parseJson.convertTo[OllamaChatResponse]
    } yield chatResponse.message.content

    F.async[String] { cb =>
      future.onComplete {
        case scala.util.Success(response) =>
          logger.info(s"Ollama chat completion succeeded for model: $actualModel")
          cb(Right(response))
        case scala.util.Failure(e) =>
          logger.warn(s"Ollama chat completion failed for model: $actualModel", e)
          cb(Left(e))
      }
    }
  }

  def textGeneration[F[_]](model: String, prompt: String)(
      implicit F: Concurrent[F]
  ): F[String] = {
    val actualModel   = if (model.isEmpty) defaultModel else model
    val promptPreview = if (prompt.length <= 250) prompt else prompt.take(250) + "..."
    logger.info(
      s"Ollama textGeneration - input model: '$model', actualModel: '$actualModel', defaultModel: '$defaultModel', prompt: '$promptPreview'"
    )
    val requestBody = OllamaGenerateRequest(
      model = actualModel,
      prompt = prompt,
      stream = false
    ).toJson.toString()

    val future: Future[String] = for {
      response <- Http()(system).singleRequest(
                   HttpRequest(
                     method = HttpMethods.POST,
                     uri = s"$baseUrl/api/generate",
                     entity = HttpEntity(ContentTypes.`application/json`, requestBody)
                   )
                 )
      responseBody     <- Unmarshal(response.entity).to[String]
      generateResponse = responseBody.parseJson.convertTo[OllamaGenerateResponse]
    } yield generateResponse.response

    F.async[String] { cb =>
      future.onComplete {
        case scala.util.Success(response) =>
          logger.info(s"Ollama text generation succeeded for model: $actualModel")
          cb(Right(response))
        case scala.util.Failure(e) =>
          logger.warn(s"Ollama text generation failed for model: $actualModel", e)
          cb(Left(e))
      }
    }
  }

  def listModels[F[_]]()(
      implicit F: Concurrent[F]
  ): F[List[String]] = {
    val future: Future[List[String]] = for {
      response <- Http()(system).singleRequest(
                   HttpRequest(
                     method = HttpMethods.GET,
                     uri = s"$baseUrl/api/tags"
                   )
                 )
      responseBody   <- Unmarshal(response.entity).to[String]
      modelsResponse = responseBody.parseJson.convertTo[OllamaModelsResponse]
    } yield modelsResponse.models.map(_.name)

    F.async[List[String]] { cb =>
      future.onComplete {
        case scala.util.Success(models) =>
          logger.info(s"Ollama listModels succeeded (${models.size} models available)")
          cb(Right(models))
        case scala.util.Failure(e) =>
          logger.warn("Ollama listModels failed", e)
          cb(Left(e))
      }
    }
  }
}

object OllamaServiceImpl {

  private[this] val logger: Logger = Logger[this.type]

  /**
    * Provides the appropriate Ollama service based on configuration and environment.
    * Priority order for enabling Ollama:
    *   1. Environment variable: OLLAMA_ENABLED = true/false
    *   2. Configuration: ollama.enabled = true/false
    *   3. Default: false (disabled for safety)
    * - If enabled, returns OllamaServiceImpl (will crash at startup if connection fails)
    * - If disabled, returns DisabledOllamaService
    */
  lazy val instance: OllamaService = {
    val isEnabled = isOllamaEnabled

    if (isEnabled) {
      logger.info("Ollama service is enabled - initializing with connection validation")
      new OllamaServiceImpl // This will crash at startup if connection fails
    } else {
      logger.info("Ollama service is disabled")
      new DisabledOllamaService
    }
  }
}

File: rholang/src/rust/interpreter/ollama_service.rs
use super::errors::InterpreterError;
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::env;
use std::sync::Arc;

#[derive(Clone, Debug)]
pub struct OllamaConfig {
    pub enabled: bool,
    pub base_url: String,
    pub model: String,
    pub timeout_sec: u64,
    pub validate_connection: bool,
}

impl OllamaConfig {
    pub fn from_env() -> Self {
        Self::from_config_values(
            false,
            "http://localhost:11434".to_string(),
            "llama4:latest".to_string(),
            30,
        )
    }

    pub fn from_config_values(
        config_enabled: bool,
        config_base_url: String,
        config_model: String,
        config_timeout_sec: u64,
    ) -> Self {
        let enabled = parse_bool_env("OLLAMA_ENABLED").unwrap_or(config_enabled);
        let base_url = env::var("OLLAMA_BASE_URL").unwrap_or(config_base_url);
        let model = env::var("OLLAMA_MODEL").unwrap_or(config_model);
        let timeout_sec = env::var("OLLAMA_TIMEOUT_SEC")
            .ok()
            .and_then(|v| v.parse().ok())
            .unwrap_or(config_timeout_sec);
        let validate_connection = enabled;

        Self {
            enabled,
            base_url,
            model,
            timeout_sec,
            validate_connection,
        }
    }

    pub fn disabled() -> Self {
        Self {
            enabled: false,
            base_url: "http://localhost:11434".to_string(),
            model: "llama4:latest".to_string(),
            timeout_sec: 30,
            validate_connection: false,
        }
    }
}

pub fn parse_bool_env(name: &str) -> Option<bool> {
    env::var(name)
        .ok()
        .and_then(|v| match v.to_lowercase().as_str() {
            "true" | "1" | "yes" | "on" => Some(true),
            "false" | "0" | "no" | "off" => Some(false),
            _ => None,
        })
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
}

#[derive(Serialize, Debug)]
struct ChatRequest {
    model: String,
    messages: Vec<ChatMessage>,
    stream: bool,
}

#[derive(Deserialize, Debug)]
struct ChatResponse {
    message: ChatMessage,
}

#[derive(Serialize, Debug)]
struct GenerateRequest {
    model: String,
    prompt: String,
    stream: bool,
}

#[derive(Deserialize, Debug)]
struct GenerateResponse {
    response: String,
}

#[derive(Deserialize, Debug)]
struct ModelInfo {
    name: String,
}

#[derive(Deserialize, Debug)]
struct ListModelsResponse {
    models: Vec<ModelInfo>,
}

#[derive(Clone)]
pub enum OllamaService {
    Real {
        client: Client,
        base_url: String,
        model: String,
        timeout_sec: u64,
    },
    Mock {
        chat_response: String,
        generate_response: String,
        models_response: Vec<String>,
    },
    Disabled,
}

impl OllamaService {
    pub fn new_real(base_url: &str, model: &str, timeout_sec: u64) -> Self {
        Self::Real {
            client: Client::builder()
                .timeout(std::time::Duration::from_secs(timeout_sec))
                .build()
                .unwrap_or_else(|_| Client::new()),
            base_url: base_url.to_string(),
            model: model.to_string(),
            timeout_sec,
        }
    }

    pub fn new_mock(
        chat_response: String,
        generate_response: String,
        models_response: Vec<String>,
    ) -> Self {
        Self::Mock {
            chat_response,
            generate_response,
            models_response,
        }
    }

    pub fn new_disabled() -> Self {
        Self::Disabled
    }

    pub fn from_config(config: &OllamaConfig) -> Self {
        if config.enabled {
            Self::new_real(&config.base_url, &config.model, config.timeout_sec)
        } else {
            Self::new_disabled()
        }
    }

    pub async fn chat(
        &self,
        model_override: Option<&str>,
        messages: Vec<ChatMessage>,
    ) -> Result<String, InterpreterError> {
        match self {
            Self::Real {
                client,
                base_url,
                model,
                ..
            } => {
                // Empty model string falls back to default (parity with Scala)
                let actual_model = match model_override {
                    Some(m) if !m.is_empty() => m.to_string(),
                    _ => model.clone(),
                };

                // Log prompt preview (truncated to 250 chars like Scala)
                let prompt_preview = if let Some(last_msg) = messages.last() {
                    if last_msg.content.len() <= 250 {
                        last_msg.content.clone()
                    } else {
                        format!("{}...", &last_msg.content[..250])
                    }
                } else {
                    String::new()
                };
                tracing::info!(
                    "Ollama chat - model: '{}', prompt: '{}'",
                    actual_model,
                    prompt_preview
                );

                let req = ChatRequest {
                    model: actual_model.clone(),
                    messages,
                    stream: false,
                };

                let res = client
                    .post(format!("{}/api/chat", base_url))
                    .json(&req)
                    .send()
                    .await
                    .map_err(|e| {
                        InterpreterError::OllamaError(format!("Ollama request failed: {}", e))
                    })?;

                if !res.status().is_success() {
                    let status = res.status();
                    let error_body = res
                        .text()
                        .await
                        .unwrap_or_else(|_| "<failed to read body>".to_string());
                    return Err(InterpreterError::OllamaError(format!(
                        "Ollama error: {} - {}",
                        status, error_body
                    )));
                }

                let body: ChatResponse = res.json().await.map_err(|e| {
                    InterpreterError::OllamaError(format!("Failed to parse response: {}", e))
                })?;

                tracing::info!(
                    "Ollama chat completion succeeded for model: {}",
                    actual_model
                );
                Ok(body.message.content)
            }
            Self::Mock { chat_response, .. } => Ok(chat_response.clone()),
            Self::Disabled => Err(InterpreterError::OllamaError(
                "Ollama service is disabled via configuration".to_string(),
            )),
        }
    }

    pub async fn generate(
        &self,
        model_override: Option<&str>,
        prompt: &str,
    ) -> Result<String, InterpreterError> {
        match self {
            Self::Real {
                client,
                base_url,
                model,
                ..
            } => {
                // Empty model string falls back to default (parity with Scala)
                let actual_model = match model_override {
                    Some(m) if !m.is_empty() => m.to_string(),
                    _ => model.clone(),
                };

                // Log prompt preview (truncated to 250 chars like Scala)
                let prompt_preview = if prompt.len() <= 250 {
                    prompt.to_string()
                } else {
                    format!("{}...", &prompt[..250])
                };
                tracing::info!(
                    "Ollama generate - model: '{}', prompt: '{}'",
                    actual_model,
                    prompt_preview
                );

                let req = GenerateRequest {
                    model: actual_model.clone(),
                    prompt: prompt.to_string(),
                    stream: false,
                };

                let res = client
                    .post(format!("{}/api/generate", base_url))
                    .json(&req)
                    .send()
                    .await
                    .map_err(|e| {
                        InterpreterError::OllamaError(format!("Ollama request failed: {}", e))
                    })?;

                if !res.status().is_success() {
                    let status = res.status();
                    let error_body = res
                        .text()
                        .await
                        .unwrap_or_else(|_| "<failed to read body>".to_string());
                    return Err(InterpreterError::OllamaError(format!(
                        "Ollama error: {} - {}",
                        status, error_body
                    )));
                }

                let body: GenerateResponse = res.json().await.map_err(|e| {
                    InterpreterError::OllamaError(format!("Failed to parse response: {}", e))
                })?;

                tracing::info!("Ollama generate succeeded for model: {}", actual_model);
                Ok(body.response)
            }
            Self::Mock {
                generate_response, ..
            } => Ok(generate_response.clone()),
            Self::Disabled => Err(InterpreterError::OllamaError(
                "Ollama service is disabled via configuration".to_string(),
            )),
        }
    }

    pub async fn list_models(&self) -> Result<Vec<String>, InterpreterError> {
        match self {
            Self::Real {
                client, base_url, ..
            } => {
                tracing::info!("Ollama list_models - fetching available models");

                let res = client
                    .get(format!("{}/api/tags", base_url))
                    .send()
                    .await
                    .map_err(|e| {
                        InterpreterError::OllamaError(format!("Ollama request failed: {}", e))
                    })?;

                if !res.status().is_success() {
                    let status = res.status();
                    let error_body = res
                        .text()
                        .await
                        .unwrap_or_else(|_| "<failed to read body>".to_string());
                    return Err(InterpreterError::OllamaError(format!(
                        "Ollama error: {} - {}",
                        status, error_body
                    )));
                }

                let body: ListModelsResponse = res.json().await.map_err(|e| {
                    InterpreterError::OllamaError(format!("Failed to parse response: {}", e))
                })?;

                let models: Vec<String> = body.models.into_iter().map(|m| m.name).collect();
                tracing::info!(
                    "Ollama list_models succeeded, found {} models",
                    models.len()
                );
                Ok(models)
            }
            Self::Mock {
                models_response, ..
            } => Ok(models_response.clone()),
            Self::Disabled => Err(InterpreterError::OllamaError(
                "Ollama service is disabled via configuration".to_string(),
            )),
        }
    }
}

pub type SharedOllamaService = Arc<tokio::sync::Mutex<OllamaService>>;

pub fn create_ollama_service(config: &OllamaConfig) -> SharedOllamaService {
    Arc::new(tokio::sync::Mutex::new(OllamaService::from_config(config)))
}

pub fn create_disabled_ollama_service() -> SharedOllamaService {
    Arc::new(tokio::sync::Mutex::new(OllamaService::new_disabled()))
}

/// Create Ollama service with connection validation (matches Scala's validateConnectionOrFail)
/// This should be used during node startup to ensure Ollama is reachable.
/// Returns an error if validate_connection is true and Ollama is unreachable.
pub async fn create_ollama_service_validated(
    config: &OllamaConfig,
) -> Result<SharedOllamaService, InterpreterError> {
    if !config.enabled {
        tracing::info!("Ollama service is disabled");
        return Ok(create_disabled_ollama_service());
    }

    let service = OllamaService::from_config(config);

    if config.validate_connection {
        tracing::info!("Validating Ollama connection to {}", config.base_url);
        // Test connection by listing models (same as Scala's validateConnectionOrFail)
        match service.list_models().await {
            Ok(models) => {
                tracing::info!(
                    "Ollama service connection validated successfully at {} ({} models available)",
                    config.base_url,
                    models.len()
                );
            }
            Err(e) => {
                return Err(InterpreterError::OllamaError(format!(
                    "Ollama service connection validation failed. Check that Ollama is running on {}: {}",
                    config.base_url, e
                )));
            }
        }
    } else {
        tracing::info!(
            "Ollama connection validation is disabled by config 'validate_connection=false'"
        );
    }

    Ok(Arc::new(tokio::sync::Mutex::new(service)))
}
